File: backend/app/domain/auth/capabilities.py
"""Capa de capacidades config-driven (capacidad -> roles), c-71 slice 2 D8.

Reemplaza el `require_roles` hardcodeado por endpoint para la accion de la
Cola de Revision:

- ``revisar_sesion``: decision TERMINAL de la sesion, en un solo acto —
  aprobar o anular. NO hay una capacidad separada para "resolver" un caso
  abierto: el owner del proyecto rechazo explicitamente el modelo de dos
  fases ("no existe el caso abierto, nunca dije que era un estado y no lo
  va a ser"; confirmado: "si, un solo paso: quien revisa decide", sin
  segunda instancia). Quien tiene `revisar_sesion` puede aprobar Y anular.

Sin framework ni infraestructura (D1): dominio puro, testeable sin DB/red.
"""

from __future__ import annotations

from app.domain.auth.roles import Rol

# capacidad -> conjunto de roles que la poseen. Dato de config, no logica.
CAPABILITY_ROLES: dict[str, frozenset[Rol]] = {
    # --- Circuito de revision humana (L2.5) ---------------------------------
    # c-76: el rol REVISOR fue eliminado; el COORDINADOR (ya presente) absorbe el
    # veredicto que tenia el revisor. Sin duplicado — el set no cambia de tamano.
    # c-78 D11: PROFESOR queda DELIBERADAMENTE afuera. Es lo que lo distingue del
    # COORDINADOR: mira la evidencia y el score, pero no emite el veredicto.
    # Quien pone la nota no decide si hubo fraude (regla dura #5).
    "revisar_sesion": frozenset({Rol.COORDINADOR, Rol.ADMIN_SISTEMA}),
    # --- Gestion academica ---------------------------------------------------
    # Alta/edicion de examenes, materias y comisiones. El DOCENTE vive aca: es su
    # trabajo. Quien juzga el fraude no edita el examen.
    # LECTURA del catalogo academico (que materias/comisiones/examenes existen y
    # cuales son "los mios"). Toda ESCRITURA cuelga de una capacidad mas acotada:
    # `gestionar_estructura` (materias/comisiones/padron), `crear_examenes` o
    # `gestionar_banco`. El tutor tiene esta y solo esta del lado academico.
    "gestionar_academico": frozenset(
        {Rol.TUTOR, Rol.PROFESOR, Rol.COORDINADOR, Rol.ADMIN_SISTEMA}
    ),
    # c-78 (E-04/E-03, D11): CREAR y editar examenes. Se separo de
    # `gestionar_academico` porque el TUTOR conserva esa (leer su catalogo,
    # inscribir, cerrar notas) pero PIERDE la creacion: armar el examen es
    # trabajo del PROFESOR, no de quien lo dicta y acompana. Sin este split,
    # sacarle la creacion al tutor le sacaba tambien las notas y las
    # inscripciones, que si son suyas.
    "crear_examenes": frozenset(
        {Rol.PROFESOR, Rol.COORDINADOR, Rol.ADMIN_SISTEMA}
    ),
    # c-78 (E-03): el BANCO de preguntas (categorias, preguntas, import XML).
    # Mismo criterio que `crear_examenes`, y por la misma razon esta separado:
    # el banco es el contenido con las respuestas correctas de toda la materia,
    # no el material de una comision.
    "gestionar_banco": frozenset(
        {Rol.PROFESOR, Rol.COORDINADOR, Rol.ADMIN_SISTEMA}
    ),
    # La ESTRUCTURA academica completa: materias, comisiones y el PADRON
    # (inscribir y desinscribir alumnos).
    #
    # Deliberadamente SIN TUTOR (decision del dueno, c-78): el tutor NO toca nada
    # de Materias y comisiones — ni crea, ni edita, ni inscribe, ni desinscribe.
    # Acompana su comision, cierra notas y supervisa; administrar el padron y la
    # grilla es de otro. El PROFESOR entra aca porque es quien arma la materia.
    #
    # OJO con el alcance: `gestionar_academico` (que el tutor SI tiene) es solo
    # LECTURA del catalogo. Toda escritura sobre materia/comision/inscripcion
    # cuelga de esta capacidad, endpoint por endpoint.
    "gestionar_estructura": frozenset(
        {Rol.PROFESOR, Rol.COORDINADOR, Rol.ADMIN_SISTEMA}
    ),
    # Ver las notas y sincronizarlas a Moodle: el docente necesita cerrar la nota
    # de su materia.
    "gestionar_notas": frozenset(
        {Rol.TUTOR, Rol.PROFESOR, Rol.COORDINADOR, Rol.ADMIN_SISTEMA}
    ),
    # Ver estadisticas institucionales agregadas (dashboard/exportables).
    # Deliberadamente SIN TUTOR: aunque son agregados sin PII, exponen el
    # rendimiento de CUALQUIER materia/comision/examen via query params sin
    # scoping por pertenencia (el docente podria pedir el resumen de una
    # comision ajena). Decision del owner: el tutor gestiona SU catalogo
    # (`gestionar_academico`/`gestionar_notas`) pero no mira estadisticas
    # institucionales — eso es coordinacion/administracion.
    # c-78: el PROFESOR entra aca (D11: conserva estadisticas). El TUTOR sigue
    # deliberadamente afuera, por la razon de arriba.
    "ver_estadisticas": frozenset(
        {Rol.PROFESOR, Rol.COORDINADOR, Rol.ADMIN_SISTEMA}
    ),
    # Asignar el docente a cargo de una comision (C-73 §9). Deliberadamente SIN
    # DOCENTE: quien queda a cargo decide quien devuelve la nota a Moodle y que
    # examenes puede tocar. Si el docente pudiera asignarse solo, la pertenencia
    # dejaria de ser un control (se auto-otorgaria el acceso que el control niega).
    "asignar_docente": frozenset(
        {Rol.COORDINADOR, Rol.ADMIN_SISTEMA}
    ),
    # --- Administracion del sistema -----------------------------------------
    # Umbrales, detectores, retencion. Deliberadamente SIN docente: la
    # configuracion define como se detecta el fraude; quien dicta la materia no
    # debe poder aflojarla para su propio examen.
    "configurar_sistema": frozenset({Rol.ADMIN_SISTEMA}),
    # Alta/baja de usuarios y asignacion de roles: solo admin del sistema.
    "gestionar_usuarios": frozenset({Rol.ADMIN_SISTEMA}),
    # Registro inmutable: lo lee quien audita, no quien opera. c-76-2: el rol
    # AUDITOR fue eliminado (nunca tuvo un endpoint real conectado a esta
    # capacidad — audit_router.py ya usaba require_roles(ADMIN_SISTEMA)
    # hardcodeado); queda exclusiva de ADMIN_SISTEMA.
    "ver_auditoria": frozenset({Rol.ADMIN_SISTEMA}),
    # --- Supervision en vivo -------------------------------------------------
    # Mirar sesiones en curso. c-76: el rol PROCTOR fue eliminado y el TUTOR entra
    # aca (supervisa a sus alumnos en vivo, acotado por comision en la capa de
    # aplicacion — Tarea 8). El TUTOR mira, pero NO decide: el VEREDICTO
    # (`revisar_sesion`) sigue SIN TUTOR (coordinador/admin, tras eliminarse
    # tambien REVISOR — c-76), preservando la separacion "quien dicta la materia
    # no juzga el fraude".
    "supervisar_vivo": frozenset(
        {Rol.TUTOR, Rol.PROFESOR, Rol.COORDINADOR, Rol.ADMIN_SISTEMA}
    ),
}


def tiene_capacidad(rol: Rol, capacidad: str) -> bool:
    """``True`` si ``rol`` esta en el conjunto de roles de ``capacidad``.

    Una capacidad no declarada en el mapa deniega por defecto (fail-closed)."""
    return rol in CAPABILITY_ROLES.get(capacidad, frozenset())

File: frontend/src/lib/capabilities.ts
/**
 * Capa de capacidades del frontend (C-71 slice 2, D8) — config-driven.
 *
 * El front SOLO oculta; el backend DECIDE (backstop server-side:
 * `require_capability`). Por eso este mapa tiene que ser una COPIA EXACTA de
 * `CAPABILITY_ROLES` en `backend/app/domain/auth/capabilities.py`: si el front
 * es más permisivo, muestra un botón que el backend rechaza con 403; si es más
 * restrictivo, esconde una acción que la persona sí podía hacer.
 *
 * `resolver_caso` (capacidad separada para el veredicto de una segunda fase)
 * DESAPARECIÓ: el modelo de dos fases fue rechazado explícitamente por el
 * owner del proyecto. `revisar_sesion` cubre TODO el acto — aprobar y anular,
 * en un solo paso — porque no hay segunda instancia que gatear aparte.
 * Al tocar este mapa, tocar el del backend en el mismo commit.
 */
import type { Rol } from "./types";

export type Capacidad =
  | "revisar_sesion"
  | "gestionar_academico"
  | "crear_examenes"
  | "gestionar_banco"
  | "gestionar_estructura"
  | "gestionar_notas"
  | "ver_estadisticas"
  | "configurar_sistema"
  | "gestionar_usuarios"
  | "ver_auditoria"
  | "supervisar_vivo";

/** capacidad → conjunto de roles que la poseen (dato de config, no lógica). */
const CAPABILITY_ROLES: Record<Capacidad, readonly Rol[]> = {
  // c-76: "revisor" eliminado; el coordinador absorbe el veredicto.
  // c-78 D11: 'profesor' queda AFUERA de revisar_sesion a propósito — es lo que
  // lo distingue del coordinador: mira la evidencia, no emite el veredicto.
  revisar_sesion: ["coordinador", "admin_sistema"],
  gestionar_academico: ["tutor", "profesor", "coordinador", "admin_sistema"],
  // c-78: CREAR exámenes y el BANCO se separaron de `gestionar_academico`. El
  // tutor conserva leer su catálogo, inscribir y cerrar notas, pero pierde la
  // creación: armar el examen es trabajo del profesor.
  crear_examenes: ["profesor", "coordinador", "admin_sistema"],
  gestionar_banco: ["profesor", "coordinador", "admin_sistema"],
  // Materias, comisiones y PADRÓN (inscribir/desinscribir). SIN tutor: el tutor
  // no toca nada de Materias y comisiones — ni crea, ni edita, ni inscribe.
  // Acompaña su comisión, cierra notas y supervisa. El profesor sí administra.
  gestionar_estructura: ["profesor", "coordinador", "admin_sistema"],
  gestionar_notas: ["tutor", "profesor", "coordinador", "admin_sistema"],
  // Deliberadamente SIN tutor: los filtros de /stats son query params libres sin
  // scoping por comisión.
  ver_estadisticas: ["profesor", "coordinador", "admin_sistema"],
  configurar_sistema: ["admin_sistema"],
  gestionar_usuarios: ["admin_sistema"],
  // c-76-2: "auditor" eliminado; queda exclusiva de admin_sistema (nunca hubo
  // un endpoint real conectado a esta capacidad para "auditor" en el backend).
  ver_auditoria: ["admin_sistema"],
  // c-76: "proctor" eliminado; el coordinador absorbe la supervisión global y
  // el tutor supervisa acotado a su comisión (D2, bloque 8) — el scoping por
  // comisión lo aplica el backend (autorizar_supervision_vivo_sobre_sesion),
  // acá solo se decide si el botón/ruta se muestra.
  supervisar_vivo: ["tutor", "profesor", "coordinador", "admin_sistema"],
};

/**
 * `true` si alguno de los `roles` posee la `capacidad`. Una capacidad no
 * declarada deniega por defecto (fail-closed), igual que el backend.
 */
export function tieneCapacidad(roles: readonly Rol[], capacidad: Capacidad): boolean {
  const permitidos = CAPABILITY_ROLES[capacidad];
  if (!permitidos) return false;
  return roles.some((r) => permitidos.includes(r));
}

/** Descripción en castellano llano de cada capacidad — se usa para mostrar
 * los permisos REALES de un rol en el formulario de alta/edición de usuarios
 * (en vez de una frase vaga de una línea por rol). */
export const CAPACIDAD_LABELS: Record<Capacidad, string> = {
  revisar_sesion: 'Revisar sesiones y decidir el veredicto (aprobar o anular)',
  gestionar_academico: 'Ver el catálogo académico (materias, comisiones y exámenes)',
  crear_examenes: 'Crear y configurar exámenes',
  gestionar_banco: 'Gestionar el banco de preguntas de la materia',
  ver_estadisticas: 'Ver las estadísticas institucionales',
  gestionar_estructura: 'Crear y editar materias y comisiones, e inscribir alumnos',
  gestionar_notas: 'Sincronizar notas con Moodle',
  configurar_sistema: 'Configurar el sistema (umbrales, detectores, retención)',
  gestionar_usuarios: 'Gestionar usuarios y roles',
  ver_auditoria: 'Ver el registro de auditoría',
  // Nombra las DOS cosas que habilita. Decía solo "Supervisar exámenes en vivo" y
  // el registro de sesiones quedaba invisible: al mirar los permisos del TUTOR no
  // aparecía por ningún lado, y parecía que no podía consultar lo que pasó en los
  // exámenes de sus comisiones. Sí puede, y siempre pudo — el backend lo acota a
  // las comisiones donde figura como tutor (`comision_tutor`).
  supervisar_vivo: 'Supervisar exámenes en vivo y ver el registro de sesiones (solo de sus comisiones)',
};

/** Módulo/dominio de cada capacidad — agrupa por la ENTIDAD sobre la que se
 * actúa (no por rol), igual al patrón de Sistema-de-Gestion-Convenios
 * (`Permiso.modulo`): agrupar por dominio hace más fácil entender de un
 * vistazo QUÉ PARTE del sistema puede tocar cada rol. */
export const CAPACIDAD_MODULO: Record<Capacidad, string> = {
  gestionar_academico: 'Académico',
  crear_examenes: 'Académico',
  gestionar_banco: 'Académico',
  gestionar_estructura: 'Académico',
  gestionar_notas: 'Académico',
  ver_estadisticas: 'Supervisión',
  supervisar_vivo: 'Supervisión',
  revisar_sesion: 'Supervisión',
  gestionar_usuarios: 'Sistema',
  configurar_sistema: 'Sistema',
  ver_auditoria: 'Sistema',
};

/** Orden fijo de módulos (no alfabético — de lo más frecuente/operativo a lo
 * más restringido/administrativo). */
const ORDEN_MODULOS = ['Académico', 'Supervisión', 'Sistema'] as const;

/** Lista de capacidades que tiene un rol (para mostrar sus permisos reales). */
export function capacidadesDeRol(rol: Rol): Capacidad[] {
  return (Object.keys(CAPABILITY_ROLES) as Capacidad[]).filter((c) =>
    CAPABILITY_ROLES[c].includes(rol)
  );
}

/** Capacidades de un rol, agrupadas por módulo/dominio y en el orden fijo de
 * `ORDEN_MODULOS` (para renderizar un bloque por módulo, patrón Convenios). */
export function permisosPorModuloDeRol(rol: Rol): { modulo: string; capacidades: Capacidad[] }[] {
  const propias = capacidadesDeRol(rol);
  return ORDEN_MODULOS
    .map((modulo) => ({
      modulo,
      capacidades: propias.filter((c) => CAPACIDAD_MODULO[c] === modulo),
    }))
    .filter((grupo) => grupo.capacidades.length > 0);
}
